### DISCUSSION TITLE
Allow to pass in `CookieOptions` when setting cookie

### DISCUSSION THREAD

**Discussion by bjarnef** [umbraco/Umbraco-CMS#14221] [Features and ideas]
Looking at the cookie managers in Umbraco core, it seems it isn't possible to pass in `CookieOptions`?

E.g. `SetCookieValue()` in `AspNetCoreCookieManager`:
https://github.com/umbraco/Umbraco-CMS/blob/dfde6798d9896e1e1bd71b9e31d9f6c5415a9d9a/src/Umbraco.Web.Common/AspNetCore/AspNetCoreCookieManager.cs#L31

or if injecting `ICookieManager`:
https://github.com/umbraco/Umbraco-CMS/blob/dfde6798d9896e1e1bd71b9e31d9f6c5415a9d9a/src/Umbraco.Core/Web/ICookieManager.cs#L9

Should it has an overload method of optional parameter to pass in `CookieOptions`?

**Comment by bergmania** [comment]
I don't like the idea of our interface exposing the `CookieOptions` as it is a third party model and our interfaces should not expose those.

I see three possible directions:
1) We make `ICookieManager` internal, as it really just is intended for us to handle cookies in the same way.
2) We expose a custom model, with a subset of the options that can be mapped to `CookieOptions`
3) We keep it as it is, and let people use it if it can benefit them and force people to handle complex cookie things them selves.

Personally I like 1) and 3) the most.

**Reply by bjarnef** [reply]
In a current project I tried used `ICookieManager` which does set/update the cookie, but always as session cookie since the expire date isn't set.

I solved it without using the `ICookieManager` instead as it didn't allow us to set expire date of the cookie.

```
if (newValue?.Any() == true)
{
    var cookieOptions = new Microsoft.AspNetCore.Http.CookieOptions()
    {
        Path = "/",
        HttpOnly = false,
        IsEssential = true,
        Expires = DateTime.Now.AddDays(1),
    };

    Context.Response.Cookies.Delete(cookieName);
    Context.Response.Cookies.Append(cookieName, string.Join(", ", newValue), cookieOptions);
}
```
